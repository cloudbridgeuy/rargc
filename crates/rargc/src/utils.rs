use convert_case::{Boundary, Converter, Pattern};

/// Convert a string to cobol case
pub fn to_cobol_case(value: &str) -> String {
    Converter::new()
        .set_pattern(Pattern::Uppercase)
        .set_delim("-")
        .set_boundaries(&[Boundary::Underscore, Boundary::LowerUpper, Boundary::Hyphen])
        .convert(value)
}

/// Convert a string from snake_case to camelCase
pub fn hyphens_to_underscores(name: &str) -> String {
    name.replace('-', "_")
}

/// Escape a string for use in a shell command
pub fn escape_shell_words(value: &str) -> String {
    let mut output = String::new();
    if value.is_empty() {
        return "''".to_string();
    }
    for ch in value.chars() {
        match ch {
            'A'..='Z' | 'a'..='z' | '0'..='9' | '_' | '-' | '.' | ',' | ':' | '/' | '@' => {
                output.push(ch)
            }
            '\n' => output.push_str("'\n'"),
            _ => {
                output.push('\\');
                output.push(ch);
            }
        }
    }
    output
}

/// Is the character a valid choice value terminator
pub fn is_choice_value_terminate(c: char) -> bool {
    c == '|' || c == ']'
}

/// Is the character a valid default value terminator
pub fn is_default_value_terminate(c: char) -> bool {
    c.is_whitespace()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cobol() {
        assert_eq!("FOO-BAR".to_string(), to_cobol_case("fooBar"));
        assert_eq!("FOO-BAR".to_string(), to_cobol_case("foo-bar"));
        assert_eq!("FOO1".to_string(), to_cobol_case("foo1"));
    }
}
