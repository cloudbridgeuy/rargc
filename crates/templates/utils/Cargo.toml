[package]
name = "utils"
version = "0.0.0"
homepage = "https://github.com/cloudbridgeuy/rargc"
description = "Tests utility library"
autobins = false

authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
name = "utils"
path = "src/lib.rs"

[dependencies]



