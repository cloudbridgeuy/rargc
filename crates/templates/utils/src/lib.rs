#[macro_export]
macro_rules! test_template {
    ($template:expr, $description:expr, $json:expr) => {
        let object = Context::from_serialize($json);

        let output =
            match TEMPLATES.render($template, &object.expect("Can't create JSON object")) {
                Ok(o) => o,
                Err(e) => {
                    log::error!("Parsing error(s): {}", e);
                    ::std::process::exit(1);
                }
            };

        insta::with_settings!({
            description => $description,
        }, {
            insta::assert_snapshot!(output)
        });
    }
}
