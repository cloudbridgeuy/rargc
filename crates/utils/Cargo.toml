[package]
name = "utils"
version = "0.0.0"
homepage = "https://github.com/cloudbridgeuy/rargs"
description = "Tests utility library"
autobins = false

authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
name = "utils"
path = "src/lib.rs"

[dependencies]
paste = "1.0.14"
