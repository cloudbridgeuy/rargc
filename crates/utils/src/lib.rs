pub extern crate paste;

#[macro_export]
macro_rules! test_template {
    ($template:expr, $description:expr, $json:expr) => {
        let object = Context::from_serialize($json);

        let output =
            match TEMPLATES.render($template, &object.expect("Can't create JSON object")) {
                Ok(o) => o,
                Err(e) => {
                    log::error!("Parsing error(s): {}", e);
                    ::std::process::exit(1);
                }
            };

        insta::with_settings!({
            description => $description,
        }, {
            insta::assert_snapshot!(output)
        });
    }
}

#[macro_export]
macro_rules! test_script {
    ($script:expr, $stream:expr, $options:expr) => {
        let script_name = $script.to_string();
        let options = $options.to_string();
        let stream = $stream.to_string();
        let description = format!("[{stream}] {script_name} {options}");
        let output_dir = "./examples/output".to_string();
        let src = format!("./examples/{script_name}");
        let dst = format!("./examples/output/{script_name}");

        log::info!("{}", description);
        let output = Command::new(&dst)
            .args(options.split(" ").collect::<Vec<&str>>())
            .output()
            .unwrap_or_else(|e| panic!("failed to execute the '{dst}' script with error: {e}"));
        insta::with_settings!({
            description => description,
        }, {
            insta::assert_snapshot!(String::from_utf8(match $stream {
                "stdout" => output.stdout,
                "stderr" => output.stderr,
                _ => panic!("invalid stream"),
            }).unwrap_or_else(|e| panic!("failed to execute the '{dst}' script with error: {e}")))
        });
    };
}

#[macro_export]
macro_rules! build_script {
    ($script:expr) => {
        let script_name = $script.to_string();
        let output_dir = "./examples/output".to_string();
        let src = format!("./examples/{script_name}");
        let dst = format!("./examples/output/{script_name}");

        log::info!("Build ./examples/output/minimal.rs");
        let output = Command::new("cargo")
            .args([
                "run",
                "--bin",
                "rargc",
                "--release",
                "--",
                "build",
                "-d",
                &output_dir,
                &src,
            ])
            .output()
            .unwrap_or_else(|e| panic!("failed to build rargc script with error: {e}"));

        println!("stdout: {}", String::from_utf8(output.stdout).unwrap());
        println!("stderr: {}", String::from_utf8(output.stderr).unwrap());

        log::info!("Check that the {dst} file exists");
        assert!(std::path::Path::new(&dst).exists());

        // Read the `dst` file into a variable of type `&str`
        let dst = std::fs::read_to_string(&dst)
            .unwrap_or_else(|e| panic!("failed to read the '{dst}' script with error: {e}"));

        insta::with_settings!({
            description => script_name,
        }, {
            insta::assert_snapshot!(dst)
        });
    };
}
