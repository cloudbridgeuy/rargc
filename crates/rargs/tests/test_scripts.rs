use std::process::Command;
use test_log::test;

use utils::{build_script, test_script};

struct Commands {
    stdout: Vec<&'static str>,
    stderr: Vec<&'static str>,
}

#[test]
fn test_minimal_sh_script() {
    let script = "minimal.sh";
    let commands = Commands {
        stdout: vec![
            "--help",
            "-h",
            "--version",
            "-v",
            "source",
            "source target",
            "source target -f",
            "source -f target",
            "--fail",
        ],
        stderr: vec!["", "-f"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_minus_v_sh_script() {
    let script = "minus-v.sh";
    let commands = Commands {
        stdout: vec!["--help", "-h", "--version", "-v"],
        stderr: vec![],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_commands_sh_script() {
    let script = "commands.sh";
    let commands = Commands {
        stdout: vec![
            "--help",
            "-h",
            "--version",
            "-v",
            "download --help",
            "download -h",
            "download example.com",
            "download example.com ./output",
            "download example.com -f",
            "download example.com ./output -f",
            "download example.com -f ./output",
            "d --help",
            "d -h",
            "d example.com",
            "d example.com ./output",
            "d example.com -f",
            "d example.com ./output -f",
            "d example.com -f ./output",
            "upload --help",
            "upload -h",
            "upload source --user Foo",
            "upload source --password Bar",
            "upload source --password Foo --password Bar",
            "u source --user Foo",
            "u source --password Bar",
            "u source --password Foo --password Bar",
        ],
        stderr: vec![
            "",
            "download",
            "d",
            "upload --user Foo",
            "upload --password Bar",
            "upload --password Foo --password Bar",
            "u --help",
            "u -h",
            "u --user Foo",
            "u --password Bar",
            "u --user Foo --password Bar",
        ],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_commands_default() {
    let script = "commands-default.sh";
    let commands = Commands {
        stdout: vec![
            "--help",
            "-h",
            "--version",
            "-v",
            "source",
            "source -f",
            "upload --help",
            "upload -h",
            "upload source",
            "upload source -f",
            "download source",
        ],
        stderr: vec!["", "-f", "upload", "upload -f"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_commands_default_force() {
    let script = "commands-default-force.sh";
    let commands = Commands {
        stdout: vec![
            "--help",
            "-h",
            "--version",
            "-v",
            "source",
            "source -f",
            "upload --help",
            "upload -h",
            "upload source",
            "upload source -f",
            "download source",
        ],
        stderr: vec!["", "-f", "upload", "upload -f"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_commands_nested() {
    let script = "commands-nested.sh";
    let commands = Commands {
        stdout: vec![
            "--help",
            "-h",
            "--version",
            "-v",
            "dir --help",
            "dir -h",
            "dir list --help",
            "dir list -h",
            "file --help",
            "file -h",
        ],
        stderr: vec!["", "dir", "file"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_command_aliases() {
    let script = "commands-aliases.sh";
    let commands = Commands {
        stdout: vec![
            "download --help",
            "d --help",
            "down --help",
            "upload --help",
            "u --help",
            "push --help",
            "push example.com",
        ],
        stderr: vec!["", "download", "upload", "d", "u", "down", "push"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_dependencies() {
    let scripts = "dependencies.sh";
    let commands = Commands {
        stdout: vec![],
        stderr: vec!["", "download", "upload"],
    };

    build_script!(scripts);

    for options in commands.stderr {
        test_script!(scripts, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(scripts, "stdout", options);
    }
}

#[test]
fn test_catch_all() {
    let script = "catch-all.sh";
    let commands = Commands {
        stdout: vec![
            "-h",
            "other",
            "required -h",
            "multiple -h",
            "no-multiple -h",
            "multiple something",
            "no-multiple something",
            "multiple something with --additional args",
            "no-multiple something with --additional args",
            "multiple something with --additional args and --debug",
            "no-multiple something with --additional args and --debug",
            "multiple something with --additional args and --debug -- with something else",
            "no-multiple something with --additional args and --debug -- with something else",
            "other something with --additional args and --debug -- with something else",
        ],
        stderr: vec!["", "required"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_catch_all_global() {
    let script = "catch-all-global.sh";
    let commands = Commands {
        stdout: vec![
            "other",
            "something, something, the dark side",
            "-h",
            "multiple -h",
            "no-multiple -h",
            "multiple something",
            "no-multiple something",
            "multiple something with --additional args",
            "no-multiple something with --additional args",
            "multiple something with --additional args and --debug",
            "no-multiple something with --additional args and --debug",
            "multiple something with --additional args and --debug -- with something else",
            "no-multiple something with --additional args and --debug -- with something else",
            "other something with --additional args and --debug -- with something else",
        ],
        stderr: vec!["", "multiple", "no-multiple"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_catch_all_stdin() {
    let script = "catch-all-stdin.sh";
    let commands = Commands {
        stdout: vec!["-h", "<<<foo"],
        stderr: vec!["--format toml"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_extensible() {
    let script = "extensible.sh";
    let commands = Commands {
        stdout: vec![
            "-h",
            "example --debug -f",
            "download -h",
            "download SOURCE",
            "upload -h",
            "upload SOURCE",
        ],
        stderr: vec!["", "upload", "download"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_extensible_delegate() {
    let script = "extensible-delegate.sh";
    let commands = Commands {
        stdout: vec![
            "-h",
            "log --pretty",
            "download -h",
            "download SOURCE",
            "upload -h",
            "upload SOURCE",
        ],
        stderr: vec!["", "upload", "download"],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_whitelist() {
    let script = "whitelist.sh";
    let commands = Commands {
        stdout: vec![
            "-h",
            "eu --user admin",
            "eu production --user admin --protocol ssh",
        ],
        stderr: vec![
            "",
            "america",
            "america --user admin",
            "eu --user hacker",
            "eu --user hacker --protocol icmp",
        ],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_repeatable_arg() {
    let script = "repeatable-arg.sh";
    let commands = Commands {
        stdout: vec![
            "-h",
            "file1",
            "file file2 file3* --action downcase",
            "file file2 file3 --action downcase --action upcase",
            "formats -h",
            "formats --action downcase",
            "formats jpg --action downcase",
            "formats jpg --action downcase gif",
            "formats jpg png --action downcase gif",
            "formats jpg --action upcase png --action downcase gif",
        ],
        stderr: vec![
            "",
            "file file2 file3 --action downcase --action --upcase",
            "formats jpg tiff --action upcase png --action downcase gif",
        ],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_repeatable_flag() {
    let script = "repeatable-flag.sh";
    let commands = Commands {
        stdout: vec![
            "",
            "-h",
            "-d 1 -d 2 -d3",
            "-d 1 -d 2 -d 3",
            "-d 1 -d 2 -d 3 -v",
            "-d 1 -d 2 -d 3 -v -v",
            "-d 1 -d 2 -d 3 -v -v -v",
            "-d 1 -d 2 -d 3 -vvv",
            "-d 1 -d 2 -d 3 -vv -v",
        ],
        stderr: vec![],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_private() {
    let script = "private.sh";
    let commands = Commands {
        stdout: vec![
            "-h",
            "connect --protocol ssh localhost",
            "connect --protocol ftp localhost",
            "connect --protocol ftp -v localhost",
            "connect --protocol ssh -v localhost",
            "connect-ssh --username foo localhost",
            "connect-ftp --username foo localhost",
        ],
        stderr: vec![
            "",
            "connect --protocol http -v",
            "connect-ftp -v --username foo localhost",
            "connect-ssh --protocol ftp --username foo localhost",
        ],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_stdin() {
    let script = "stdin.sh";
    let commands = Commands {
        stdout: vec!["-h", "arg --help", "option --help"],
        stderr: vec![],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_examples() {
    let script = "examples.sh";
    let commands = Commands {
        stdout: vec!["-h"],
        stderr: vec![],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_empty() {
    let script = "empty.sh";
    let commands = Commands {
        stdout: vec!["-v", "--version", "-help", "-h"],
        stderr: vec![""],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        println!("stdout: {}", options);
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_hello_world() {
    let script = "hello-world.sh";
    let commands = Commands {
        stdout: vec!["", "-v", "--version", "-help", "-h"],
        stderr: vec![],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_huggingface() {
    let script = "huggingface.sh";
    let commands = Commands {
        stdout: vec!["", "-v", "--version", "-help", "-h"],
        stderr: vec![],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_first_option_help() {
    let script = "first-option-help.sh";
    let commands = Commands {
        stdout: vec![
            "",
            "-h",
            "--help",
            "download -h",
            "download --help",
            "upload -h",
            "upload --help",
        ],
        stderr: vec![
            "-v -h",
            "-v --help",
            "download source -h",
            "download source --help",
            "upload source -h",
            "upload source --help",
        ],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        println!("stdout: {}", options);
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_no_first_option_help_global() {
    let script = "no-first-option-help-global.sh";
    let commands = Commands {
        stdout: vec![
            "",
            "-h",
            "--help",
            "download -h",
            "download --help",
            "upload -h",
            "upload --help",
            "-v -h",
            "-v --help",
            "download source -h",
            "download source --help",
            "upload source -h",
            "upload source --help",
        ],
        stderr: vec![],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_no_first_option_help_local() {
    let script = "no-first-option-help-local.sh";
    let commands = Commands {
        stdout: vec![
            "",
            "-h",
            "--help",
            "download -h",
            "download --help",
            "upload -h",
            "upload --help",
            "download source -h",
            "download source --help",
        ],
        stderr: vec![
            "-v -h",
            "-v --help",
            "upload source -h",
            "upload source --help",
        ],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_minimal_with_comments() {
    let script = "minimal-with-comments.sh";
    let commands = Commands {
        stdout: vec!["-h", "--help"],
        stderr: vec![""],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_json() {
    let script = "json.sh";
    let commands = Commands {
        stdout: vec!["", "-h", "--help"],
        stderr: vec![],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}

#[test]
fn test_flags() {
    let script = "flags.sh";
    let commands = Commands {
        stdout: vec![
            "--falsy",
            "--no-truthy",
            "--truthy",
            "--shorty",
            "--no-shorty",
            "-no-s",
            "-h",
            "--help",
            "",
        ],
        stderr: vec![
            "--falsy",
            "--no-truthy",
            "--truthy",
            "--shorty",
            "--no-shorty",
            "-no-s",
            "",
        ],
    };

    build_script!(script);

    for options in commands.stderr {
        test_script!(script, "stderr", options);
    }

    for options in commands.stdout {
        test_script!(script, "stdout", options);
    }
}
