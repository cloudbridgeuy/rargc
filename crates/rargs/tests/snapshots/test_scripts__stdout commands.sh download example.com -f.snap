---
source: crates/rargs/tests/test_scripts.rs
description: "[stdout] commands.sh download example.com -f"
expression: "String::from_utf8(match \"stdout\" {\n            \"stdout\" => output.stdout,\n            \"stderr\" => output.stderr,\n            _ => panic!(\"invalid stream\"),\n        }).unwrap_or_else(|e|\n        panic!(\"failed to execute the '{dst}' script with error: {e}\"))"
---

args:
- rargs_force = 1
- rargs_input = ([0] = "example.com"
- [1] = "-f")
- rargs_source = example.com

